text/0000-discriminant-intrinsic.md

```diff
@@ -0,0 +1,339 @@
+- Start Date: 2015-01-21
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+Add a new intrinsic, `discriminant_value` that extracts the value of the discriminant for enum
+types.
+
+# Motivation
+
+Many operations that work with discriminant values can be significantly improved with the ability to
+extract the value of the discriminant that is used to distinguish between variants in an enum. While
+trivial cases often optimise well, more complex ones would benefit from direct access to this value.
+
+A good example is the `SqlState` enum from the `postgres` crate (Listed at the end of this RFC). It
+contains 233 variants, of which all but one contain no fields. The most obvious implementation of
+(for example) the `PartialEq` trait looks like this:
+
+```rust
+match (self, other) {
+    (&Unknown(ref s1), &Unknown(ref s2)) => s1 == s2,
+    (&SuccessfulCompletion, &SuccessfulCompletion) => true,
+    (&Warning, &Warning) => true,
+    (&DynamicResultSetsReturned, &DynamicResultSetsReturned) => true,
+    (&ImplicitZeroBitPadding, &ImplicitZeroBitPadding) => true,
+	      .
+		  .
+		  .
+    (_, _) => false
+}
+```
+
+Even with optimisations enabled, this code is very suboptimal, producing
+[this code](https://gist.github.com/Aatch/c23a45634b10aaecad05). A way to extract the discriminant
+would allow this code:
+
+```rust
+match (self, other) {
+    (&Unknown(ref s1), &Unknown(ref s2)) => s1 == s2,
+    (l, r) => unsafe {
+	    discriminant_value(l) == discriminant_value(r)
+	}
+}
+```
+
+Which is compiled into [this IR](https://gist.github.com/Aatch/beb736b93a908aa67e84).
+
+# Detailed design
+
+## What is a discriminant?
+
+A discriminant is a value stored in an enum type that indicates which variant the value is. The most
+common case is that the discriminant is stored directly as an extra field in the variant. However,
+the discriminant may be stored in any place, and in any format. However, we can always extract the
+discriminant from the value somehow.
+
+## Implementation
+
+For any given type, `discriminant_value` will return a `u64` value. The values returned are as
+specified:
+
+* **Non-Enum Type**: Always 0
+* **Enum Type**: A value that uniquely identifies that variant within its type. I.E. for a given
+  enum typem `E`, and two values of type `E`, `a` and `b`, the expression `discriminant_value(a) ==
+  discriminant_value(b)` is true iff `a` and `b` are the same variant. Two values of different types
```

Comment:
This definition precludes the use of `discriminant_value` for `derive(PartialOrd)` or a form of inequality between variants. That may be fine for now, just want to confirm that decision.